### ISSUE TITLE
Bump targetSdkVersion to Android 10 and adapt to behavior changes

### ISSUE THREAD

**Issue by tsapeta** [expo/expo#8469]
**Edit: We're working on this and it turns out to be a change on the bigger side. This does not affect apps that are already in the Play Store until November. If you're working on an app that hasn't been submitted to the Play Store, you'll either need to submit it before August 3 (according to Google's docs) or wait a week or so till we're able to build support for this into SDK 38.**

---

Followup #8450 where we bumped only `compileSdkVersion` and left `targetSdkVersion` being set to 28. The decision not to bump `targetSdkVersion` was made due to some behavior changes that would potentially hit us, especially https://developer.android.com/reference/android/Manifest.permission#ACCESS_BACKGROUND_LOCATION
We also have to check whether there are no other behavior changes after bumping `targetSdkVersion` to 29.

EDIT:
Google starts to send this message while uploading app to google play:
![Screenshot_2020-07-23_14-10-30 (1)](https://user-images.githubusercontent.com/47157550/88507036-23e62580-cfe4-11ea-83ca-a7e3da19b79b.png)

There are multiple threads about this on expo forum:
https://forums.expo.io/t/from-august-2020-new-apps-must-target-at-least-android-10/40699
https://forums.expo.io/t/anforderungen-an-die-ziel-api-ebene-ab-august-2020/40827
Google [documentation](https://support.google.com/googleplay/android-developer/answer/113469#targetsdk) says that 'From August 2020' means 'August 3, 2020':

<img src='https://i.ibb.co/TK7c7DY/screenshot-418.png'  height='500'/>

**Comment by midsummerit** [comment]
Can I asked for an update on this please. I uploaded a app to the Google Play Store today (31st July) and I get this warning.

"Currently, your app's target API level is 28 and should be at least 29 to ensure it compiles with the latest APIs, which offer optimal security and performance.

As of August 2020, new apps must set at least target API level 29 (Android 10).
Starting in November 2020, app updates should set target API level 29 (Android 10)."

I am on SDK version 38 and using the Expo build server. Google *will not* let me move to production with the error outstanding.

**Comment by adrynalyne** [comment]
I would be interested in an update as well, seeing how we are bumping up against the deadline.

**Comment by ide** [comment]
We're working on this and it turns out to be a change on the bigger side. This does not affect apps that are already in the Play Store until November. If you're working on an app that hasn't been submitted to the Play Store, you'll either need to submit it before August 3 (according to Google's docs) or wait a week or so till we're able to build support for this into SDK 38.

**Comment by SleeplessByte** [comment]
(Hack: update an AAB/APK to a closed track asap, it will count as update as soon as there is one in there somewehre)

**Comment by Mamtak** [comment]
My expo app target to sdkVersion 38 and i am uploading before august 3 but still its not letting me to upload the Build on Play store.

**Comment by brentvatne** [comment]
@Mamtak that is only a warning, it will work

**Comment by midsummerit** [comment]
It does not let me proceed to release. Maybe I am missing something. Can
you tell us how you did a release even with the API version warning.

TIA

**Comment by Troy-Harder** [comment]
Hello, I cannot upload my app using the latest expo version due to API 28. Where are you gonna fix that or How to upload an app? Thanks in advance for your response.

**Comment by ide** [comment]
We are working on this updating the Android for SDK 38 this week.

**Comment by peterclink** [comment]
@ide any updates?

**Comment by ide** [comment]
As of August 5th, 14:38 UTC the SDK 38 Android builders will produce AABs/APKs with the targetSdkVersion set to 29 (Android 10).

* Apps built before that moment will remain on targetSdkVersion = 28.
* You will need to rebuild your app after that moment for your AAB/APK to include this change.
* If you're using turtle-cli to build your standalone apps, to have this change included in your build, please upgrade to version equal or higher than 0.17.1.

**Comment by josmas** [comment]
@ide Any advice on how to upgrade for bare workflow? I am looking at the [template](https://github.com/expo/expo/blob/sdk-38/templates/expo-template-bare-minimum/android/build.gradle) and it has not yet been updated, and we generally compare against that when upgrading.

I have bumped our build tools, compile and target sdk, and support lib all to 29 and it seems to be working, but I want to be cautious (expo version is 38.0.9).
I can only see https://github.com/expo/expo/pull/8450 as the main source of the change, but would appreciate any pointers to other changes I should review, if any.

**Comment by ide** [comment]
Here are the other Android 10 PRs I'm aware of:

- https://github.com/expo/expo/pull/9446 (main one)
- https://github.com/expo/expo/pull/9514
- https://github.com/expo/expo/pull/9666

39 should be coming in about a month too (exact date unknown, sometime before the end of September) if you'd just rather wait.

**Comment by josmas** [comment]
Thanks a million @ide, that's very helpful. Will go through them.

I'd love to wait but our users cannot publish any new apps they create in our platform, so we have to find a way around it.

**Comment by lc3t35** [comment]
Hi, what is the status on this issue : how can we build an ejected SDK38 app with api level 29 support ?

Please, this is URGENT.

**Comment by ide** [comment]
You will need to follow Google's instructions for targeting API Level 29 and then if there are individual libraries that have issues (see the posts above for some light direction) you will need to upgrade those.

**Comment by lc3t35** [comment]
Hi @ide , you wrote August 3rd : "We are working on updating the Android for SDK 38 for Android 10 support this week."
Please give clear instructions.

**Comment by josmas** [comment]
You can update as any other Android app, bumping up your tools, target SDK and dependencies and making sure all libraries you use are either compatible or not affected by the changes in 29 (specified as [behaviour changes](https://developer.android.com/about/versions/10/behavior-changes-10)).
We have a number of apps in the store built with this upgrade and haven't had any issues so far.

**Comment by lc3t35** [comment]
Thank you @josmas, but I need to know if Expo's team detected any problem with sdk38 and Android 10 ...
This is what I was expecting by "We are working on updating the Android for SDK 38 for Android 10 support this week.", if they update there is a problem or not ?

**Comment by ide** [comment]
@lc3t35 we've updated the latest code on the master branch to work with Android 10, backported the changes to the sdk-38 branch, and updated the managed builds (`expo build:android`) to produce AABs that target Android 10. We are now focused on SDK 39+.

Here are the PRs I know of that were needed for Android 10 compatibility for SDK 38:

https://github.com/expo/expo/pull/9446
https://github.com/expo/expo/pull/9514
https://github.com/expo/expo/pull/9666
https://github.com/expo/expo/pull/8450

**Comment by lc3t35** [comment]
thank you @ide , my app is an ejected sdk38 app, did you test this case when you backport ? what is the recommended process to update existing ejected sdk38 app ?

**Comment by ide** [comment]
Ejected apps were not explicitly tested. The code in ejected apps is similar to that of managed apps that are built with the Expo build service so I imagine things should largely work.

There's no official recommended process but as mentioned above I'd follow Google's guidelines and use the latest versions of the packages that are compatible with SDK 38 (namely the ones touched in the PRs above) and go through your regular QA steps. 

In short, there are Android 10-compatible versions of Expo-authored packages on npm and the rest is standard Android programming. Alternatively you can wait for SDK 39 which will be available sometime this month.

**Comment by lc3t35** [comment]
Thank you very much James !
